backend/de.metas.business/src/main/java/de/metas/request/api/impl/RequestDAO.java

```diff
@@ -71,36 +82,50 @@ public I_R_Request createRequest(@NonNull final RequestCandidate candidate)
-		request.setC_Order_ID(candidate.getRecordRef() != null ?
-				candidate.getRecordRef().getTableName().equals(I_C_Order.Table_Name) ? candidate.getRecordRef().getRecord_ID() : -1
-				: -1);
-		request.setM_InOut_ID(candidate.getRecordRef() != null ?
-				candidate.getRecordRef().getTableName().equals(I_M_InOut.Table_Name) ? candidate.getRecordRef().getRecord_ID() : -1
-				: -1);
-
-		request.setDateTrx(SystemTime.asTimestamp());
+		optionalOfFirstNonNullSupplied(candidate::getOrderId,
+				() -> getIdFromReferenceOrNull(candidate, I_C_Order.Table_Name, OrderId::ofRepoId))
+				.ifPresent(orderId -> request.setC_Order_ID(orderId.getRepoId()));
+		optionalOfFirstNonNullSupplied(candidate::getInOutId,
+				() -> getIdFromReferenceOrNull(candidate, I_M_InOut.Table_Name, InOutId::ofRepoId))
+				.ifPresent(inOutId -> request.setM_InOut_ID(inOutId.getRepoId()));
+		optionalOfFirstNonNullSupplied(candidate::getInvoiceId,
+				() -> getIdFromReferenceOrNull(candidate, I_C_Invoice.Table_Name, InvoiceId::ofRepoId))
+				.ifPresent(invoiceId -> request.setM_InOut_ID(invoiceId.getRepoId()));
+
+		optionalOfFirstNonNullSupplied(candidate::getPaymentId,
+				() -> getIdFromReferenceOrNull(candidate, I_C_Invoice.Table_Name, PaymentId::ofRepoId))
+				.ifPresent(paymentId -> request.setM_InOut_ID(paymentId.getRepoId()));
+
+		request.setC_Project_ID(ProjectId.toRepoId(candidate.getProjectId()));
+
+		request.setDateTrx(coalesce(asTimestamp(candidate.getDateTrx()), SystemTime.asTimestamp()));
+		request.setReminderDate(asTimestamp(candidate.getReminderDate()));
 		request.setAD_User_ID(UserId.toRepoId(candidate.getUserId()));
 		request.setR_RequestType_ID(candidate.getRequestTypeId().getRepoId());
 		request.setM_QualityNote_ID(QualityNoteId.toRepoId(candidate.getQualityNoteId()));
 		request.setPerformanceType(candidate.getPerformanceType());
-		request.setDateDelivered(TimeUtil.asTimestamp(candidate.getDateDelivered()));
-		request.setResult(candidate.getResult());
+		request.setDateDelivered(asTimestamp(candidate.getDateDelivered()));
 		request.setR_Status_ID(RequestStatusId.toRepoId(candidate.getStatusId()));
-		request.setPriority(RequestPriority.toValue(candidate.getPriority()));
-		if (candidate.getIsEscalated() != null)
-		{
-			request.setIsEscalated(candidate.getIsEscalated());
-		}
-		if (candidate.getIsSelfService() != null)
+		if (candidate.getPriority() != null)
 		{
-			request.setIsSelfService(candidate.getIsSelfService());
+			request.setPriority(RequestPriority.toValue(candidate.getPriority()));
 		}
+		request.setSalesRep_ID(UserId.toRepoId(candidate.getSalesRepId()));
+		request.setC_BP_Vendor_ID(BPartnerId.toRepoId(candidate.getVendorId()));
 
 		save(request);
 
 		return request;
 	}
 
+	@Nullable
+	private static <T extends RepoIdAware> T getIdFromReferenceOrNull(final @NonNull RequestCandidate candidate, @NonNull final String tableName, @NonNull final Function<Integer, T> idMapper)
+	{
+		return candidate.getRecordRef() != null ?
```

Comment:
The nested ternary operators (?:) make the logic dense and difficult to read in a single line.

Reply:
refactored to the more traditional if formula.